### PR TITLE
Add Sphinx documentation generation foundation

### PR DESCRIPTION
**Pull request by HyukjinKwon**
This PR proposes to add a foundation for the documentation and API list for databricks.koalas (a.k.a Spark Pandas). This PR

1. Add `Makefile` and Sphinx configuration files to generate the documentation from docstrings.

2. Fix `.travis.xml` and `dev/lint-python` to test documentation generation.

3. Add `__version__` to `databricks.koalas.version` that is shared for `setup.py`, `conf.py` (for doc generation) and `databricks.koalas.__version__`. This hack is borrowed from PySpark approach.

4. Add `requirements-dev.txt` file so that it can be used in CI and dev guide.

5. Add some extra additional packages to generate docs that need to be installed:
    - `nbsphinx` to allow Jupyter note books. Seems needing some more works to actually use it.
    - `numpydoc` to allow `numpy` style docs (from Pandas via `derived_from`).

Note and followups:
 - It is important to address #90 before finishing the documentation generation as it decides how we document (see https://github.com/databricks/spark-pandas/pull/84#pullrequestreview-228161984)

  - We should also discuss how we're going to deal with `derived_from`. See https://github.com/databricks/spark-pandas/pull/84#discussion_r276867315

Partially addresses #87

### CODE REVIEW

File: databricks/koala/__init__.py [hunk 1]
```diff
@@ -14,6 +14,8 @@
 # limitations under the License.
 #
 
+__version__ = "0.0.5"
```
Comment (thunterdb) [inline databricks/koala/__init__.py]: Is there a way to pull this value into `setup.py`? Otherwile they will get out of sync

File: docs/conf.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import os
+import sys
+from databricks import koala
+sys.path.insert(0, os.path.abspath('.'))
+
+
+# -- Project information -----------------------------------------------------
+
+project = 'databricks-koala'
+copyright = '2019, Timothy Hunter'
```
Comment (thunterdb) [inline docs/conf.py]: Databricks is the copyright holder.

File: databricks/koala/frame.py [hunk 3]
```diff
@@ -59,12 +58,10 @@ def _index_columns(self):
         return [anchor_wrap(self, self._spark_getitem(field))
                 for field in self._metadata.index_fields]
 
-    @derived_from(pd.DataFrame)
```
Comment (thunterdb) [inline databricks/koala/frame.py]: Let's keep them, because they add documentation from the pandas Dataframe automatically. I just tried and here is what it looks like:

Screenshot 2019-04-18 at 10 52 25

File: docs/conf.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+# -- General configuration ---------------------------------------------------
+
+# If your documentation needs a minimal Sphinx version, state it here.
+needs_sphinx = '1.2'
+
+# Add any Sphinx extension module names here, as strings. They can be
+# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
+# ones.
+extensions = [
+    'sphinx.ext.autodoc',
```
Comment (thunterdb) [inline docs/conf.py]: Do you know if there is an extension for rendering jupyter notebooks? I would like to add jupyter notebooks, see #34 . This can be done separately if this is too much effort.

File: docs/databricks.koala.rst [hunk 5]
```diff
@@ -0,0 +1,51 @@
+databricks.koala package
+==========================
```
Comment (thunterdb) [inline docs/databricks.koala.rst]: As mentioned, we do not need to document the koala package, but only:
 - SparkSession: https://github.com/databricks/spark-pandas/blob/master/databricks/koala/utils.py#L63
 - DataFrame: https://github.com/databricks/spark-pandas/blob/master/databricks/koala/utils.py#L52
 - Column: https://github.com/databricks/spark-pandas/blob/master/databricks/koala/utils.py#L53

So `pyspark.sql.dataframe` etc. should be the only 3 required modules.

**Comment by thunterdb** [comment]
Oh yes I had to add that locally on conda.

**Comment by HyukjinKwon** [comment]
Thanks guys. Will address comments soon.

File: databricks/koala/frame.py (hunk 3 above)
Comment (HyukjinKwon) [reply databricks/koala/frame.py]: Hm, I removed this because the doctests were wrongly copied too. But seems they are supposed to be skipped. Let me take a look.

Comment (HyukjinKwon) [reply databricks/koala/frame.py]: Hm, let me handle this separately.

File: databricks/koala/__init__.py (hunk 1 above)
Comment (HyukjinKwon) [reply databricks/koala/__init__.py]: Possible but it's a bit hacky:

https://github.com/apache/spark/blob/5cdb8a23df6f269d6be0bf3536e9af9e29c4a05f/python/setup.py#L31-L37

Let me try.

File: docs/conf.py (hunk 4 above)
Comment (HyukjinKwon) [reply docs/conf.py]: Yes, seems we can refer https://nbsphinx.readthedocs.io/en/0.4.2/. Let me add the plugin accordingly for now.

File: .travis.yml [hunk 6]
```diff
@@ -44,13 +44,14 @@ install:
   - conda info -a
 
   # Replace dep1 dep2 ... with your dependencies
-  - conda create -c conda-forge -q -n test-environment python=$TRAVIS_PYTHON_VERSION nose==1.3.7 flake8>=3.5.0 pandas==$PANDAS_VERSION pyarrow==$PYARROW_VERSION decorator
+  - conda create -c conda-forge -q -n test-environment python=$TRAVIS_PYTHON_VERSION sphinx==2.0.1 nbsphinx=0.4.2 numpydoc=0.8.0 nose==1.3.7 flake8>=3.5.0 pandas==$PANDAS_VERSION pyarrow==$PYARROW_VERSION decorator
```
Comment (HyukjinKwon) [inline .travis.yml]: note: we gonna make a requirement.txt for dev

File: docs/conf.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    'sphinx.ext.viewcode',
+    'numpydoc',  # handle NumPy documentation formatted docstrings. Needs to install
+    'nbsphinx',  # Jupyter Notebook. Needs to install
```
Comment (HyukjinKwon) [inline docs/conf.py]: both plugins should be installed for dev

**Comment by HyukjinKwon** [comment]
@rxin, I fixed the PR description since this PR doesn't fully address #87. For current status, this PR will only add the foundation for documentation generation.

**Comment by HyukjinKwon** [comment]
I believe it's ready to go ..

File: dev/lint-python [hunk 8]
```diff
@@ -27,6 +27,12 @@ MINIMUM_PYCODESTYLE="2.4.0"
 
 SPHINX_BUILD="sphinx-build"
 
+# Koalas depends on PySpark
```
Comment (rxin) [inline dev/lint-python]: why is this change needed? (lint depends on Spark?)

File: docs/conf.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import sys
+from databricks import koalas
+sys.path.insert(0, os.path.abspath('.'))
+
+
+# -- Project information -----------------------------------------------------
+
+project = 'databricks-koalas'
+copyright = '2019, Databricks'
+author = 'Timothy Hunter'
```
Comment (rxin) [inline docs/conf.py]: I'd change this to "The Koalas Team"

File: docs/conf.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+import os
+import sys
+from databricks import koalas
+sys.path.insert(0, os.path.abspath('.'))
+
+
+# -- Project information -----------------------------------------------------
+
+project = 'databricks-koalas'
```
Comment (rxin) [inline docs/conf.py]: koalas?

File: docs/index.rst [hunk 11]
```diff
@@ -0,0 +1,17 @@
+.. databricks-koalas documentation master file, created by
+   sphinx-quickstart on Thu Apr 18 14:08:15 2019.
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+Welcome to databricks-koalas's documentation!
```
Comment (rxin) [inline docs/index.rst]: Welcome to Koalas' documentation!

**Comment by rxin** [comment]
Can you also update README to add a small section on building docs?

File: .travis.yml (hunk 6 above)
Comment (rxin) [reply .travis.yml]: just update the README.md file

**Comment by rxin** [comment]
Also do you expect anything to show up? I built the doc against master but it's mostly empty at the moment.

File: dev/lint-python (hunk 8 above)
Comment (HyukjinKwon) [reply dev/lint-python]: Doc generation actually imports the package to peek attributes.

Currently, Koalas checks if Pyspark is importable or not when Koalas is imported.

**Comment by HyukjinKwon** [comment]
rst files should be updated later when we finish encapsulation. Then the documentation will be properly shown.

Will address other comments later tonight (KST)

File: .travis.yml (hunk 6 above)
Comment (HyukjinKwon) [reply .travis.yml]: Will do. But `requirement.txt` for dev is still needed for adding other dependencies like `nose` . Otherwise, we should keep adding dependecies in this line.

Pandas has a separate file for dev too https://github.com/pandas-dev/pandas/blob/master/requirements-dev.txt

Comment (rxin) [reply .travis.yml]: make sense. is there a way for conda to pick up that file as well? then we don't need that readme instruction to list requirements and instead just have a conda command.

Comment (HyukjinKwon) [reply .travis.yml]: Seems there are two ways:

 - To have `requirements.txt` for PIP and `environment.yml` for Conda (with conversion script [generate_pip_deps_from_conda.py](https://github.com/pandas-dev/pandas/blob/022f458ee858721d366c27dd88940d69383ab960/scripts/generate_pip_deps_from_conda.py)). Seems what Pandas does.

- Looks we are able to share requirements.txt to use it in Conda too (for instance see https://www.technologyscout.net/2017/11/how-to-install-dependencies-from-a-requirements-txt-file-with-conda/).

I will make `environment.yml` for now, and take a separate look later for this one.

File: .travis.yml [hunk 12]
```diff
@@ -14,7 +14,7 @@ matrix:
     - python: "3.5"
       env:
         - SPARK_VERSION=2.3.3
-        - PANDAS_VERSION=0.19.2
+        - PANDAS_VERSION=0.23.0
```
Comment (HyukjinKwon) [inline .travis.yml]: Looks like we don't have to test against Pandas 0.19.2:

https://github.com/databricks/spark-pandas/blob/44ff66babdc30dabbdecbf6a140da9ff045d9e94/setup.py#L48

File: README.md [hunk 13]
```diff
@@ -92,7 +88,7 @@ We recommend setting up a Conda environment for development:
 ```bash
 conda create --name koalas-dev-env python=3.6
 conda activate koalas-dev-env
-conda install -c conda-forge pyspark=2.4 pandas pyarrow=0.10 decorator flake8 nose
+conda install -c conda-forge --yes -n test-environment --file requirements-dev.txt
```
Comment (ueshin) [inline README.md]: `-n test-environment` is not necessary?

**Comment by HyukjinKwon** [comment]
I believe this is ready for a look.

File: README.md [hunk 14]
```diff
@@ -62,10 +62,6 @@ df.columns = ['x', 'y', 'z1']
 df['x2'] = df.x * df.x
 ```
 
-## Documentation
```
Comment (HyukjinKwon) [inline README.md]: Let me leave the readthedoc link somewhere else later if you guys are fine with it.

Comment (rxin) [reply README.md]: i think we should still keep this until we have published docs. Otherwise it looks like we created a project without any docs and even without the realization we need docs.

**Comment by rxin** [comment]
Looks great. Feel free to merge.

File: README.md (hunk 14 above)
Comment (HyukjinKwon) [reply README.md]: Okie. Let me keep it

**Comment by HyukjinKwon** [comment]
Thanks, @rxin. Can anyone merge this? I don't have the access yet so I can't merge.

**Comment by rxin** [comment]
Try again - I just added you.

**Comment by HyukjinKwon** [comment]
Thanks, @rxin.
